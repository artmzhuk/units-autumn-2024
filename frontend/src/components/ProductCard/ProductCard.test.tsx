import {render, screen} from '@testing-library/react';
import {ProductCard} from '../ProductCard';
import {getPrice} from '../../utils';
import '@testing-library/jest-dom';

jest.mock('../../utils', () => ({
    getPrice: jest.fn(),
}));

describe('ProductCard Component', () => {
    const product = {
        id: 1,
        name: 'IPhone 14 Pro',
        description: 'Latest iPhone model',
        price: 999,
        category: 'Электроника',
        imgUrl: '/iphone.png',
    };

    beforeEach(() => {
        (getPrice as jest.Mock).mockImplementation((price, symbol) => `${price} ${symbol}`);
    });

    it('should render the product details correctly', () => {
        render(<ProductCard id={1}
                            name={'IPhone 14 Pro'}
                            description={'Latest iPhone model'}
                            price={999}
                            priceSymbol={'$'}
                            category={'Электроника'}
                            imgUrl={'/iphone.png'}/>);
        expect(screen.getByText('IPhone 14 Pro')).toBeInTheDocument();
        expect(screen.getByText('Latest iPhone model')).toBeInTheDocument();
        expect(screen.getByText('Электроника')).toBeInTheDocument();
        expect(screen.getByText('999 $')).toBeInTheDocument(); // price formatted by getPrice
    });

    it('should render the product image when imgUrl is provided', () => {
        render(<ProductCard id={1}
                            name={'IPhone 14 Pro'}
                            description={'Latest iPhone model'}
                            price={999}
                            priceSymbol={'$'}
                            category={'Электроника'}
                            imgUrl={'/iphone.png'}/>);
        const image = screen.getByAltText('IPhone 14 Pro');
        expect(image).toBeInTheDocument();
        expect(image).toHaveAttribute('src', '/iphone.png');
    });

    it('should not render the product image when imgUrl is missing', () => {

        render(<ProductCard id={1}
                            name={'IPhone 14 Pro'}
                            description={'Latest iPhone model'}
                            price={999}
                            priceSymbol={'$'}
                            category={'Электроника'}
        />);
        expect(screen.queryByAltText('IPhone 14 Pro')).not.toBeInTheDocument();
    });

        it('snapshot render without image', () => {
        const rendered = render(
            <ProductCard
                name="IPhone 14 Pro"
                description="Latest iPhone model"
                price={999}
                priceSymbol="$"
                category="Электроника"
                id={2}
            />
        );

        expect(rendered.asFragment()).toMatchSnapshot();
    });
});
