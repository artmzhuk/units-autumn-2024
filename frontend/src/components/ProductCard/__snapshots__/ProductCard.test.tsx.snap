// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Component snapshot render without image 1`] = `
<DocumentFragment>
  <div
    class="product-card"
  >
    <div
      class="product-card__text"
    >
      <h2
        class="product-card__name"
      >
        IPhone 14 Pro
      </h2>
      <div
        class="product-card__description"
      >
        Latest iPhone model
      </div>
      <h1
        class="product-card__price"
      >
        999 $
      </h1>
      <div
        class="product-card__category"
      >
        Электроника
      </div>
    </div>
  </div>
</DocumentFragment>
`;
